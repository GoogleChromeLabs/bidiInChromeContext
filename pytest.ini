# https://docs.pytest.org/en/latest/reference/customize.html
[pytest]
minversion = 7.0
testpaths = tests

# pytest-md-report plug-in
md_report = True
