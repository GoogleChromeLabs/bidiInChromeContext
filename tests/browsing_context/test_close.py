# Copyright 2023 Google LLC.
# Copyright (c) Microsoft Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import pytest
from syrupy.filters import paths
from test_helpers import (execute_command, get_tree, goto_url,
                          read_JSON_message, send_JSON_command, subscribe,
                          wait_for_event)


@pytest.mark.asyncio
async def test_browsingContext_close(websocket, context_id):
    await subscribe(websocket, ["browsingContext.contextDestroyed"])

    command_id = await send_JSON_command(websocket, {
        "method": "browsingContext.close",
        "params": {
            "context": context_id
        }
    })

    # Assert "browsingContext.contextCreated" event emitted.
    resp = await read_JSON_message(websocket)
    assert resp == {
        'type': 'event',
        "method": "browsingContext.contextDestroyed",
        "params": {
            "context": context_id,
            "parent": None,
            "url": "about:blank",
            "children": None,
            "userContext": "default"
        }
    }

    resp = await read_JSON_message(websocket)
    assert resp == {"type": "success", "id": command_id, "result": {}}

    result = await get_tree(websocket)

    # Assert context is closed.
    assert result == {'contexts': []}


@pytest.mark.asyncio
async def test_browsingContext_close_prompt(websocket, context_id, html,
                                            snapshot):
    await subscribe(websocket, [
        "browsingContext.userPromptOpened", "browsingContext.contextDestroyed"
    ])

    url = html("""
        <script>
            window.addEventListener('beforeunload', event => {
                event.returnValue = 'Leave?';
                event.preventDefault();
            });
        </script>
        """)

    await goto_url(websocket, context_id, url)

    # We need to interact with the page to trigger "beforeunload"
    result = await execute_command(
        websocket, {
            "method": "script.evaluate",
            "params": {
                "expression": "document.body.click()",
                "target": {
                    "context": context_id
                },
                "awaitPromise": False,
                "userActivation": True
            }
        })

    command_id = await send_JSON_command(
        websocket, {
            "method": "browsingContext.close",
            "params": {
                "context": context_id,
                "promptUnload": True
            }
        })

    # Assert "browsingContext.userPromptOpened" event emitted.
    resp = await read_JSON_message(websocket)
    assert resp == {
        'type': 'event',
        "method": "browsingContext.userPromptOpened",
        "params": {
            "context": context_id,
            "message": "",
            "type": "beforeunload"
        }
    }

    await send_JSON_command(
        websocket, {
            "method": "browsingContext.handleUserPrompt",
            "params": {
                "context": context_id,
            }
        })

    # Assert "browsingContext.contextDestroyed"" event emitted.
    response = await wait_for_event(websocket,
                                    "browsingContext.contextDestroyed")
    assert response == snapshot(exclude=paths("params.context"))
    assert response['params']['context'] == context_id

    resp = await read_JSON_message(websocket)
    assert resp == {"type": "success", "id": command_id, "result": {}}

    result = await get_tree(websocket)

    # Assert context is closed.
    assert result == {'contexts': []}


@pytest.mark.asyncio
async def test_browsingContext_close_viaUserContext(websocket):
    user_context = await execute_command(websocket, {
        "method": "browser.createUserContext",
        "params": {}
    })

    await subscribe(websocket, ["browsingContext.contextDestroyed"])

    browsing_context = await execute_command(
        websocket, {
            "method": "browsingContext.create",
            "params": {
                "type": "window",
                "userContext": user_context["userContext"]
            }
        })

    tree = await execute_command(websocket, {
        "method": "browsingContext.getTree",
        "params": {}
    })

    assert len(tree['contexts']) == 2

    await send_JSON_command(websocket, {
        "method": "browser.removeUserContext",
        "params": user_context
    })

    destroyed = await wait_for_event(websocket,
                                     "browsingContext.contextDestroyed")

    assert destroyed['params']['context'] == browsing_context["context"]

    tree = await execute_command(websocket, {
        "method": "browsingContext.getTree",
        "params": {}
    })

    assert len(tree['contexts']) == 1
    assert tree["contexts"][0]["userContext"] == "default"
