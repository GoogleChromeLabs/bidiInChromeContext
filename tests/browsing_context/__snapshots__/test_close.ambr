# serializer version: 1
# name: test_browsingContext_close_prompt[websocket0]
  dict({
    'method': 'browsingContext.contextDestroyed',
    'params': dict({
      'children': None,
      'parent': None,
      'url': "data:text/html,%0A        <script>%0A            window.addEventListener('beforeunload', event => {%0A                event.returnValue = 'Leave?%27;%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20event.preventDefault();%0A%20%20%20%20%20%20%20%20%20%20%20%20});%0A%20%20%20%20%20%20%20%20%3C/script%3E",
      'userContext': 'default',
    }),
    'type': 'event',
  })
# ---
